core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java

```diff
@@ -86,32 +86,30 @@ public void onNewClusterState(ClusterState state) {
 
             @Override
             public void onClusterServiceClose() {
-                logger.warn("[{}] cluster service closed while waiting for write consistency number of shards " +
-                                "to be allocated on index creation.", indexName);
-                listener.onResponse(new ClusterStateUpdateResponse(false));
+                logger.debug("[{}] cluster service closed while waiting for enough shards to be started.", indexName);
+                listener.onFailure(new NodeClosedException(clusterService.localNode()));
             }
 
             @Override
             public void onTimeout(TimeValue timeout) {
-                logger.warn("[{}] index was created, but the operation timed out waiting for write consistency " +
-                                "number of shards to be allocated.", indexName);
-                listener.onResponse(new ClusterStateUpdateResponse(false));
+                logger.debug("[{}] index created, but the operation timed out while waiting for enough shards to be started.", indexName);
+                listener.onFailure(new ElasticsearchTimeoutException("[{}] index created, but the operation timed out while " +
+                                                                     "waiting for enough shards to be started.", indexName));
             }
         };
 
-        if (observedState.status() == ClusterState.ClusterStateStatus.APPLIED
-                && enoughShardsActive(observedState, indexName, settings, waitOnActiveShards)) {
-            observerListener.onNewClusterState(observedState);
-        } else {
-            observer.waitForNextChange(observerListener, shardsAllocatedPredicate, timeout);
-        }
+        observer.waitForNextChange(observerListener, shardsAllocatedPredicate, timeout);
     }
 
     // package private for testing
     static boolean enoughShardsActive(final ClusterState clusterState, final String indexName,
```

Comment:
I wonder if we should move this method (and checkWaitForActiveShards in ReplicationOperation) directly into ActiveShardCount